### ISSUE TITLE
STJ generator with external (generated) context class

### ISSUE THREAD

**Issue by alrz** [dotnet/runtime#108317]
Currently, you cannot stack Roslyn source generators to operate on generated code. As a workaround I opted into running my own generators in a separate project. That is,  `[JsonSerializable(...)] class MyContext : JsonSerializerContext` is generated elsewhere, however this requires the STJ generator to recognize external context classes.

Project1:
```cs
// use a custom generator to generate the context, 
// note this is only a "blueprint" since STJ generator is not yet run.

// <auto-genreated/>
[JsonSerializable(...)] 
[JsonSerializable(...)] 
public class GeneratedContext : JsonSerializerContex;
```

Project2 references Project1:
```cs
// STJ generator won't recognize GeneratedContext in scope
// but perhaps we can use something like this to hint on that:

[JsonSerializerExternalContext(typeof(GeneratedContext))]
partial class MyContext : JsonSerializerContex;
```
This way we can stack generators using separate projects but the workaround only works if the final generator can work with external types.

**Comment by elgonzo** [comment]
Have you already tried using [`JsonTypeInfoResolver.Combine`](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.serialization.metadata.jsontypeinforesolver.combine) to combine the `MyContext.Default` and `GeneratedContext.Default` resolvers?

**Comment by alrz** [comment]
> GeneratedContext.Default

That doesn't exist, note when you generate the context STJ generator won't see it. I did elaborate why this is the case in OP.

**Comment by alrz** [comment]
>What motivates this use case?

It's a workaround for stacking generators that roslyn doesn't support. The scenario is being able to generate the context itself.

>It's something we had considered, but it didn't quite seem to justify the added complexity

Since something like `JsonSerializerExternalContext` (or maybe [MetadataTypeAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations.metadatatypeattribute?view=net-8.0) since the usage is similar) is a direct hint for the generator to know where to look at, I don't think it's necessarily adds too much complexity or perf hint for this to be considered?

**Comment by eiriktsarpalis** [comment]
Oh I see, the context on the first project is generated by a custom generator which you want to take into account in the second project. It seems like a niche scenario if I'm honest, it's very unlikely we would be able to prioritize it in the near future.

**Comment by eiriktsarpalis** [comment]
It's not that simple. You need to propose a new attribute for including context. There's also design challenges. What happens if you specify two contexts and they have conflicting contracts for a given type? What happens if you have a `]JsonSerializable(typeof(Foo))]` in the last context and the transitive closure of either of the earlier contexts also includes a conflicting definition for `Foo`? This is not an easy problem.

**Comment by alrz** [comment]
> conflicting definition 

Isn't all that already possible to happen? We're just adding another source for attributes (the "metadata class"), the semantics of the sum of the applied attributes remain intact. 

This definitely needs a design review pass but in any case I can start a prototype for that to happen at some point. For one, I'd like this attribute to be sufficient for the generator to find the type so there's practically two entries, JsonSerializableAttribute and MetadataTypeAttribute (or similar).

**Comment by dferretti** [comment]
Same issue here - any source generator that generates any type might want those types to have source generated serializer contexts.

The general problem is like OP says, source generators (like STJ) cannot reference outputs of other source generators.

I have not dug too deep in the STJ generator yet, but would it be possible for STJ to expose its serializer context generation as a library method? Then my source generator could generate the same context classes that STJ would have generated, by referencing STJ as a source generator / compile time dependency.

**Comment by gregsdennis** [comment]
> It seems like a niche scenario if I'm honest, it's very unlikely we would be able to prioritize it in the near future. - @eiriktsarpalis 

I don't think this is a niche scenario.  As mentioned above, a very common scenario is API client generation, e.g. from an OpenAPI description.  Perhaps more niche is my intent to eventually support generating API (server) code from OpenAPI.

**Comment by eiriktsarpalis** [comment]
> I don't think this is a niche scenario. As mentioned above, a very common scenario is API client generation, e.g. from an OpenAPI description. Perhaps more niche is my intent to eventually support generating API (server) code from OpenAPI.

Could you share more details? What motivates the use of stacking contexts in that scenario over just doing an E2E generation at the application layer? If the concern is application size I believe the linker will do a good job at trimming away contexts generated at the library layer that are not being used.

We generally recommend doing E2E contract generation at the application layer using the final type graphs. In practice stacking contexts (e.g. as is currently already achievable using `TypeInfoResolverChain`) creates more problems than it solves, especially when generic types come into play.

**Comment by gregsdennis** [comment]
Consider a nuget package that generates API client code.  It provides an attribute that you put on a partial class. 

```c#
[ApiClient("https://someserver.com/api/openapi.yml")]
public partial class SomeApiClient;
```

This generates all the methods and models needed to interact with the API.

As the developer of the nuget package, it makes sense to want to add those models to a serializer context.
